utils/estimated_histogram.hh

```diff
@@ -110,214 +111,7 @@ namespace utils {
-
-    /*
-     * \brief returns the count in the given bucket
-     */
-    uint64_t get(size_t bucket) const {
-        return _buckets[bucket];
-    }
-
-    /*!
-     * \brief get a histogram quantile
-     *
-     * This method will returns the estimated value at a given quantile.
-     * If there are N values in the histogram.
-     * It would look for the bucket that the total number of elements in the buckets
-     * before it are less than N * quantile and return that bucket lower limit.
-     *
-     * For example, quantile(0.5) will find the bucket that that sum of all buckets values
-     * below it is less than half and will return that bucket lower limit.
-     * In this example, this is a median estimation.
-     *
-     * It will return 0 if the histogram is empty.
-     *
-     */
-    uint64_t quantile(float quantile) const {
-        if (quantile < 0 || quantile > 1.0) {
-            throw std::runtime_error("Invalid quantile value " + std::to_string(quantile) + ". Value should be between 0 and 1");
-        }
-        auto c = count();
-
-        if (!c) {
-            return 0; // no data
-        }
-
-        auto pcount = uint64_t(std::floor(c * quantile));
-        uint64_t elements = 0;
-        for (size_t i = 0; i < NUM_BUCKETS - 2; i++) {
-            if (_buckets[i]) {
-                elements += _buckets[i];
-                if (elements >= pcount) {
-                    return get_bucket_lower_limit(i);
-                }
-            }
-        }
-        return Max; // overflowed value is in the requested quantile
-    }
-
-    /*!
-     * \brief returns the mean histogram value (average of bucket offsets, weighted by count)
-     * It will return 0 if the histogram is empty.
-     */
-    uint64_t mean() const {
-        uint64_t elements = 0;
-        double sum = 0;
-        for (size_t i = 0; i < NUM_BUCKETS - 1; i++) {
-            elements += _buckets[i];
-            sum += _buckets[i] * get_bucket_lower_limit(i);
-        }
-        return (elements) ?  sum / elements : 0;
-    }
-
-    /*!
-     * \brief returns the number of buckets;
-     */
-    size_t size() const {
-        return NUM_BUCKETS;
-    }
-
-    /*!
-     * \brief returns the total number of values inserted
-     */
-    uint64_t count() const {
-        uint64_t sum = 0L;
-        for (size_t i = 0; i < NUM_BUCKETS; i++) {
-            sum += _buckets[i];
-        }
-        return sum;
-    }
-
-    /*!
-     * \brief multiple all the buckets content in the histogram by a constant
-     */
-    approx_exponential_histogram& operator*=(double v) {
-        for (size_t i = 0; i < NUM_BUCKETS; i++) {
-            _buckets[i] *= v;
-        }
-        return *this;
-    }
-
-    uint64_t& operator[](size_t b) noexcept {
-        return _buckets[b];
-    }
-};
+using approx_exponential_histogram = seastar::metrics::internal::approximate_exponential_histogram<Min, Max, Precision>;
```

Comment:
We should not be using `seastar::*::internal` types in ScyllaDB code. Code in the `seastar::*::internal` namespace is subject to change in backwards-incompatible ways. Even if this code is very unlikely to change in any way, it is good to stick to this principle.

If you think applications should start using `seastar::metrics::internal::approximate_exponential_histogram<>`, I suggest you send a patch to seastar first, to make it public, by moving it out of the `internal` namespace. Note that this might require documenting that class (if it isn't yet) as we require public seastar types to be documented.

Reply:
Here's the deal: this code was in Scylla for a few years. A few months back, I needed it in Seastar, so Seastar got its copy of the estimated histogram. For Seastar, it's internal. As you said, it may change in the future (it's actually going to, see below)

Now we have the same code in two places. In both places, it's used for the same purpose, histograms for the monitoring layer.

The code in this patch uses aliasing so that ScyllaDB will have the same classes as before, but it will remove the code duplication.

Besides code de-duplication, I'm adding Prometheus Native-histogram support, which would require more information from the histogram that would require changing the code in two places.

Not the end of the world, but it could be avoided

Reply:
I'm all for getting rid of code duplication, I'm asking for making these types public in seastar first.